### PR TITLE
Update final-form.md

### PR DESCRIPTION
**Pull request by yoshisatoyanagisawa**
Add effectiveConnectionType condition.
With this condition, web developers can decide using ServiceWorker or not upon the network quality.

### CODE REVIEW

**Comment by yoshisatoyanagisawa** [comment]
@sisidovski @quasi-mod Will you take a look?
Although https://wicg.github.io/netinfo/ is WICG draft written 4 years ago, the condition is usable to enable ServiceWorker only for low quality networks.

File: final-form.md [hunk 1]
```diff
@@ -239,3 +244,16 @@ addEventListener('install', (event) => {
...
 });
 ```
+
+### Avoid using ServiceWorker for 4g network 
+```js
+// If network connection is good enough, use network directly.
+addEventListener('install', (event) => {
+  event.addRoutes({
+    condition: {
+      effectiveConnectionType: "4g"
```
Comment (sisidovski) [inline final-form.md:254]: How can we write a rule for slow networks i.e. worse than 3g? Should we encourage to use `or`?

**Review by sisidovski — COMMENTED** [review]
Probably it's good to ensure if ECT is really effective or not before updating the explainer.

File: final-form.md (hunk 1 above)
Comment (yoshisatoyanagisawa) [reply final-form.md:254]: I encourage people to explicitly use `or` for that case.
Therefore, "worse than 3g" can be written like: 
```javascript
  not: {
    or: [
      {effectiveConnectionType: "4g"},
      {effectiveConnectionType: "3g"}
    ]}
```

or
```javascript
  or: [
    {effectiveConnectionType: "slow-2g"},
    {effectiveConnectionType: "2g"}
  ]
```

**Comment by yoshisatoyanagisawa** [comment]
@domenic Let me ask your thoughts on the condition.

**Comment by domenic** [comment]
I guess the network information API is Chrome-only so this will not get much support from other vendors. We might want to discuss this a bit in the explainer.

IIUC, if another vendor does not implement this, then the condition will be seen as `{}`. Can we ensure that the spec does not throw an error in that case, and just ignores the condition instead?

I think the `or` usage is a reasonable way to do conjunctions. I think the version with just `or` is simpler than the version with `or` + `not` so I would encourage including an example like that.

**Comment by tunetheweb** [comment]
BTW the general consensus is that ECT is no longer fit for purpose as on most sites as nearly all traffic is 4G these days.

Taking the BBC as a an example of widely used service across the world:
- [On mobile](https://lookerstudio.google.com/c/u/0/reporting/bbc5698d-57bb-4969-9e07-68810b9fa348/page/TvuS?params=%7B%22origin%22:%22www.bbc.com%22,%22df72%22:%22include%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580phone%22%7D) it is ~86% 4G, 13% 3G and basically nothing for the rest.
- [On desktop](https://lookerstudio.google.com/c/u/0/reporting/bbc5698d-57bb-4969-9e07-68810b9fa348/page/TvuS?params=%7B%22origin%22:%22www.bbc.com%22,%22df72%22:%22include%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580desktop%22%7D) it's 100% 4G.

More and more people are giving up on ECT. There has also been a few lack-luster attempts to add new categories but no consensus and seems unlikely to happen.

So is it worth adding this feature? If we do want to do this, maybe we should look at [the RTT property of the NetworkInformation API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt) which is more future proof than the crude categories in ECT.

**Comment by yoshisatoyanagisawa** [comment]
> I guess the network information API is Chrome-only so this will not get much support from other vendors. We might want to discuss this a bit in the explainer.

Sure.

> IIUC, if another vendor does not implement this, then the condition will be seen as `{}`. Can we ensure that the spec does not throw an error in that case, and just ignores the condition instead?

Ah, I did not think about the case but it sounds we need to think/decide before expanding the spec.
In case of a new condition or a new source, there might be three options.

Option 1. raise if the condition is unknown.  The web author should write the rule that works for both. If we choose this route, addRoutes must only ignore the rules raised and add anything else.
e.g. addRoutes(ruleA) => success, addRoutes(ruleB) => raise, addRoutes(ruleC) => success, the actual rule will be ruleA + ruleC.

Option 2. ignores an unknown condition or an unknown source, but evaluate as if it the condition does not exist.
e.g. {condition: {unknown: foo, urlPattern: "/*"}, source: "network"}, it will be recorded like {condition: {urlPattern: "/*", source: "network"}.
e.g. {condition: {urlPattern: "/*"}, source: "unknown"} then raise because empty source is not allowed?

Option 3. ignores the rule with an unknown condition or an unknown source.
e.g. [{condition: {urlPattern: "/*"}, source: "network"}, {condition: {unknown: rule}, source: "network"}, {condition: {urlPattern: "/*"}, source: "unknown"} then only [{condition: {ur](condition: {urlPattern: "/*"}, source: "network"}) will be added.

Option 4. provides the way to let web developers to know supported conditions and sources.  Web developers should check the supported conditions before using it.
e.g. console.log(event.supportedConditions) then prints {"urlPattern", "requestMode", ...}

> I think the `or` usage is a reasonable way to do conjunctions. I think the version with just `or` is simpler than the version with `or` + `not` so I would encourage including an example like that.

sure.  However, there is another discussion on effectiveness of effective connection type.  Let me address that after we have consensus on ECT (or other network information field).

**Comment by yoshisatoyanagisawa** [comment]
> More and more people are giving up on ECT. There has also been a few lack-luster attempts to add new categories but no consensus and seems unlikely to happen.

Thanks for helping me to know what is happening for ECT.

> So is it worth adding this feature? If we do want to do this, maybe we should look at [the RTT property of the NetworkInformation API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt) which is more future proof than the crude categories in ECT.

Let me explain my motivation to add the field.  If the goal is fulfilled, using ECT is not important.

I would like to address people with terrible network connectivity.  I am thinking about cellular in extremely cloudy train station, airplane wifi, extremely cloudy conference room, or bad radio situation for some reasons.  In that case, I expect ServiceWorker to reduce network traffics or to provide offline capabilities.  At the same time, if people move from such terrible network to good "4g" network, I do not want ServiceWorker to bother the communication. i.e. behave as if not exist even for network performance level.

I see ECT as the way to detect such bad connection, and ServiceWorker static routing condition uses it to make the fetch handler shows up when it is needed.  For good "4g", I do not want ServiceWorker to bother the communication. i.e. no interception.  Therefore, I might need a rule like if not 4g, then fetch handler.

What do you think?

**Comment by tunetheweb** [comment]
Yeah, it is still good for "am I on a really, really bad network?" use case. But we'll see less and less people on that over time.. But there will always be a cohort like that, so maybe we've peaked?

**Comment by yoshisatoyanagisawa** [comment]
For the way to expand conditions/sources, which is pointed by @domenic in https://github.com/WICG/service-worker-static-routing-api/pull/27#issuecomment-2179956595, I have filed https://github.com/WICG/service-worker-static-routing-api/issues/28.  Let's discuss there.
(Note that the example has been updated to use "or")

> Yeah, it is still good for "am I on a really, really bad network?" use case. But we'll see less and less people on that over time.. But there will always be a cohort like that, so maybe we've peaked?

Considering average users, I agree that the number of people with bad network connection should continuously reduce.  However, I am talking about long tail bad connectivity users.  What can be provided for them?

**Comment by tunetheweb** [comment]
As mentioned, [the RTT property of the NetworkInformation API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt) allows whatever granularity you want rather than being bound by [the strict boundaries of ECT](https://wicg.github.io/netinfo/#effective-connection-types) and little desire to improve upon then. Therefore as speeds improve, you can adjust your settings (maybe start with a 270ms RTT for now to match "3g" and change to 200ms later, and then 100ms, as speeds improve?).

ECT is currently only based on RTT anyway. As per spec it should be based on RTT and download measures, but the only implementation (Chromium) actually [removed download criteria](https://chromium-review.googlesource.com/c/chromium/src/+/1469649) so ECT is currently only measured by RTT anyway.

The major downside of this is it's easier to understand "3G" compared to "270ms RTT".

**Comment by yoshisatoyanagisawa** [comment]
Sure.  I am flexible to go with RTT.  Do you have a suggested condition name?
Is condition name "rttGt", "rttGe", "rttLt", and "rttLe" make sense? e.g. "rttGt": 25, "rttLe": 500.
How about "rtt": {"lt": 500, "gt": 25}?
Or, are there any other condition names you come up with?

Also, what do you think of `downlink` ?  It is also used for deciding ECT.
Do you think it is nice to be specified with the condition?

**Comment by yoshisatoyanagisawa** [comment]
PTAL.
I have updated the spec to use `rtt: {lt: 150}` style.

**Comment by yoshisatoyanagisawa** [comment]
PTAL.
I have updated the proposed condition to use `rtt: {lt: 150}` style.

**Review by domenic — COMMENTED** [review]
I don't like these names very much. Possible alternatives:

- `lessThan`, `lessThanOrEqualTo`, `greaterThan`, `greaterThanOrEqualTo`
- `below`, `atMost`, `above`, `atLeast`

Other possible simplifications to combine with the above:

- Include only >= and <=, or only > and <, instead of both.
- Avoid nested dictionaries, so something like `rttLessThan`, `rttGreaterThan`.

**Comment by yoshisatoyanagisawa** [comment]
> I don't like these names very much. Possible alternatives:
> 
> * `lessThan`, `lessThanOrEqualTo`, `greaterThan`, `greaterThanOrEqualTo`
> * `below`, `atMost`, `above`, `atLeast`
> 
> Other possible simplifications to combine with the above:
> 
> * Include only >= and <=, or only > and <, instead of both.
> * Avoid nested dictionaries, so something like `rttLessThan`, `rttGreaterThan`.

Thanks for the suggestions.
I have updated the names to `rttLessThan` and `rttGreaterThan`.

**Review by domenic — APPROVED** [review]
LGTM

**Comment by yoshisatoyanagisawa** [comment]
@domenic Thanks for the review.
@sisidovski Will you take another look?

**Review by sisidovski — APPROVED** [review]
LGTM.

Probably it's better to clarify that `lessThan` and `greaterThan` mean `<=` or `<` more explicitly in the example. IMO the combination with `not` condition makes the example complicated, but that's nice to have.

**Comment by yoshisatoyanagisawa** [comment]
Ok, let me add an example without "not".

**Pull request merged**